crates/pop-wallet-integration/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+#[derive(Serialize, Debug, PartialEq)]
+pub enum DataType {
+	/// Parachain call, where Vec<u8> is the encoded call data.
+	Parachain(Vec<u8>),
+	Contract(ContractArgs),
+}
+
+/// Contract specific data variations.
+pub mod contracts {
+	use super::Serialize;
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct ContractArgs {
+		call_type: ContractCallType,
+		storage_deposit_limit: Option<String>,
+		/// The binary (wasm, polkavm) of the contract.
+		code: Option<Vec<u8>>,
+	}
+
+	#[derive(Serialize, Debug, PartialEq)]
+	pub enum ContractCallType {
+		// no unique fields.
+		Upload,
+		Instantiate(InstantiateArgs),
+		Call(CallArgs),
+	}
+
+	/// Arguments for instantiating a contract.
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct InstantiateArgs {
+		constructor: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+		salt: Option<Vec<u8>>,
+	}
+
+	/// Arguments for calling a contract.
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct CallArgs {
+		address: String,
+		message: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+	}
+
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+		use crate::*;
+		use serde_json::json;
+
+		#[test]
+		fn json_serialize_parachain_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Parachain(vec![0, 1, 2, 3]),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Parachain": [0,1,2,3]}})
+			);
+		}
+		#[test]
+		fn json_serialize_contract_upload_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Contract(ContractArgs {
+					call_type: ContractCallType::Upload,
+					storage_deposit_limit: Some("1234".to_string()),
+					code: Some(vec![0, 1, 2]),
+				}),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Contract": {
+					"call_type": "Upload",
+					"storage_deposit_limit": "1234",
+					"code": [0,1,2]
+				}}})
+			);
+		}
+
+		// TODO: after high level review, complete serialization tests.
+	}
+}
+use crate::contracts::ContractArgs;
+
+struct StateHandler {
+	shutdown_tx: Option<oneshot::Sender<()>>,
+	signed_payload: Option<String>,
+}
+
+/// Manages the wallet integration for secure signing of transactions.
+pub struct WalletIntegrationManager {
+	frontend_path: PathBuf,
+	// cloning can be expensive (e.g. contract code stored in-memory)
+	data: Arc<Data>,
+	signed_payload: Option<String>,
+}
+
+impl WalletIntegrationManager {
+	/// - frontend_path: Path to the wallet-integration frontend.
+	/// - data: Data to be sent to the frontend for signing.
+	pub fn new(frontend_path: PathBuf, data: Data) -> Self {
```

Comment:
Why a path? My assumption was that the UI would be bundled with the binary, whereas this implies it has to be written to some local path first. Could it not just serve the static content from a string, or is the problem the additional assets (images/css)? I would have assumed those would be better server from a CDN to keep the binary size down.

Reply:
Recommendation is to source the frontend from a GH release, similar to what we do with nodes. So, the plan is to have a path to the pop cache for the frontend.

Reply:
however, will attempt to get everything into one string still -- a good idea

Reply:
It would be nice to ship with something that works out of the box, even if just a fallback. The GH release does provide a means to push updates if the UI needs bumps (e.g. security patches), but it does arguably add complexity in caching and checking for new versions.

Reply:
Added a simple frontend trait that serves a custom route. Allows us to easily plug in different frontend sources (baked-in string, cache, etc.).

Please see: https://github.com/r0gue-io/pop-cli/pull/362/commits/0edfbad00685263f61e690b1bc0e21434352fc83